File: PYTHON/CODE/Finite.py
NO_OF_CHARS = 256

def getNextState(pat, M, state, x):
    if state < M and x == ord(pat[state]):
        return state + 1

    i = 0
    for ns in range(state, 0, -1):
        if ord(pat[ns - 1]) == x:
            while i < ns - 1:
                if pat[i] != pat[state - ns + 1 + i]:
                    break
                i += 1
            if i == ns - 1:
                return ns
    return 0

def computeTF(pat, M):
    TF = [[0 for i in range(NO_OF_CHARS)] for _ in range(M + 1)]
    for state in range(M + 1):
        for x in range(NO_OF_CHARS):
            z = getNextState(pat, M, state, x)
            TF[state][x] = z
    return TF

def search(pat, txt):
    M = len(pat)
    N = len(txt)
    TF = computeTF(pat, M) 
    state = 0
    pattern_found = False
    for i in range(N):
        state = TF[state][ord(txt[i])]
        if state == M:
            print("Pattern found at index: {}".format(i - M + 1))
            pattern_found = True
    if not pattern_found:
        print("Pattern not found")

def main():
    file_path = input("Enter the file path: ")
    try:
        with open(file_path, 'r') as file:
            lines = file.readlines()
            txt = lines[0].strip()  # First line is text
            pat = lines[1].strip()  # Second line is pattern
            search(pat, txt)
    except FileNotFoundError:
        print("File not found.")

if __name__ == '__main__':
    main()

File: C++ IMPLEMENTATION/CODE/Finite.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>

#define NO_OF_CHARS 256

int getNextState(const std::string& pat, int M, int state, char x) {
    if (state < M && x == pat[state]) {
        return state + 1;
    }

    int i = 0;
    for (int ns = state; ns > 0; ns--) {
        if (pat[ns - 1] == x) {
            while (i < ns - 1) {
                if (pat[i] != pat[state - ns + 1 + i]) {
                    break;
                }
                i++;
            }
            if (i == ns - 1) {
                return ns;
            }
        }
    }
    return 0;
}

void computeTF(const std::string& pat, int M, std::vector<std::vector<int>>& TF) {
    for (int state = 0; state <= M; state++) {
        for (int x = 0; x < NO_OF_CHARS; x++) {
            int z = getNextState(pat, M, state, x);
            TF[state][x] = z;
        }
    }
}

void search(const std::string& pat, const std::string& txt) {
    int M = pat.length();
    int N = txt.length();
    std::vector<std::vector<int>> TF(M + 1, std::vector<int>(NO_OF_CHARS));
    computeTF(pat, M, TF);

    int state = 0;
    int pattern_found = 0;

    for (int i = 0; i < N; i++) {
        state = TF[state][txt[i]];
        if (state == M) {
            std::cout << "Pattern found at index: " << i - M + 1 << std::endl;
            pattern_found = 1;
        }
    }
    if (!pattern_found) {
        std::cout << "Pattern not found" << std::endl;
    }
}

int main() {
    std::string file_path;
    std::cout << "Enter the file path: ";
    std::cin >> file_path;

    std::ifstream file(file_path);
    if (!file.is_open()) {
        std::cout << "File not found." << std::endl;
        return 1;
    }

    std::string txt, pat;

    // Read text and pattern from file
    std::getline(file, txt);  // First line is text
    std::getline(file, pat);  // Second line is pattern

    // Removing newline characters if present
    if (!txt.empty() && txt.back() == '\n') {
        txt.pop_back();
    }
    if (!pat.empty() && pat.back() == '\n') {
        pat.pop_back();
    }

    search(pat, txt);

    file.close();
    return 0;
}
